### DISCUSSION TITLE
How to allow HEAD requests?

### DISCUSSION THREAD

**Discussion by cinderblockgames** [authelia/authelia#4998] [Q&A] [answered]
The Jump dashboard uses HEAD requests to check if a site is up, but making a HEAD call to Authelia behind Traefik is resulting in a 405 Method Not Allowed.  Is there a way to enable HEAD requests?

**Comment by james-d-elliott** [comment, marked as answer]
We handle HEAD requests appropriately for most legitimate use cases, for endpoints that do not have any specific HEAD responses we appropriately respond with a 405 per web standards. What endpoint are you trying to make HEAD requests to or what are you trying to do?

**Comment by cinderblockgames** [comment]
@james-d-elliott The request is for **https://login.example.com/?rd=https%3A%2F%2Frecipes.example.com%2F&rm=HEAD** but I get a 405 just going straight to Authelia as well:

```
curl -I https://login.example.com

HTTP/1.1 405 Method Not Allowed
content-length: 22
content-type: text/plain; charset=utf-8
date: Mon, 27 Feb 2023 21:37:46 GMT
strict-transport-security: max-age=60;
```

**Reply by james-d-elliott** [reply]
That's correct, we don't support HEAD requests on the root URL. The request should be a GET request. Supporting HEAD requests here would not be beneficial, HEAD requests can only return headers, this page does nothing unless the content of the page is sent. Why is it making a HEAD request to that URL?

**Reply by cinderblockgames** [reply]
So, I made the root URL request to test.  The bolded one is the one that's being made by the application, which is also receiving a 405 Method Not Allowed.

**Reply by cinderblockgames** [reply]
(Although, the dashboard *would* request the root if I configured it as a link on the page.  The dashboard makes HEAD requests to make sure that the application is up to display a status.)

**Reply by james-d-elliott** [reply]
That's correct.. why is it making a HEAD request instead of a GET request?

**Reply by cinderblockgames** [reply]
Because it's just checking to make sure that the app is up for a status.

https://github.com/daledavies/jump/blob/main/jumpapp/classes/Status.php#L53

**Reply by james-d-elliott** [reply]
Which site is it checking?

**Reply by cinderblockgames** [reply]
So, in this instance, it's checking the recipes site, following the redirect, and getting a 405.

I also have a link to **https://login.example.com/reset-password/step1**, which is working with HEAD.  If I make that a link to **https://login.example.com**, which would also make sense in this context, I get the 405 (and thus it shows as if the site is down).

**Reply by james-d-elliott** [reply]
Then us adjusting our behavior will not help in this situation. If it's checking `https://recipes.example.com`, and we respond with a 200, it will hide the fact `https://recipes.example.com` is down when it's down. 

Jump needs to include credentials in the request in order to authenticate and not be redirected.

**Reply by cinderblockgames** [reply]
I accept that, for the recipes situation.  Stinks for my use case, but sure.

But that doesn't address the second situation, where I am providing a link to Authelia itself.  That would *also* show as down, when it's not.

**Reply by cinderblockgames** [reply]
Also.. how would Jump provide the credentials?  Do you mean it would need to automate seeing the redirect, submitting the Authelia form, and following the redirect with the provided cookie?  Or is there a simpler way?  Could I just do `https://testuser:password@recipes.example.com` and have Authelia pick that up?

**Reply by cinderblockgames** [reply]
> If it's checking `https://recipes.example.com`, and we respond with a 200, it will hide the fact `https://recipes.example.com` is down when it's down.

Also, it's at least up enough to redirect.

**Reply by james-d-elliott** [reply]
How does it stink for your use case? It will always show as up if Authelia and the proxy are up, so there's no point in having the status check at all for individual apps, it's just an inefficient request. Configuring it to be disabled is the most logical solution that I can see, if Jump doesn't allow this then that's clearly a necessary adjustment on their end as well as fixing the bug where the Cookie/Authorization headers are not being passed from the client to the backend.

As far as the second situation no one has had that use case yet including yourself as far as I can tell. We can look at doing this. But why wouldn't a status check be more configurable? For example you should be able to set request method, headers, and post content.

**Reply by james-d-elliott** [reply]
> Also, it's at least up enough to redirect.

This is an invalid argument. The proxy may arbitrarily redirect regardless of the backend services status.

**Reply by james-d-elliott** [reply]
> Also.. how would Jump provide the credentials? Do you mean it would need to automate seeing the redirect, submitting the Authelia form, and following the redirect with the provided cookie? Or is there a simpler way? Could I just do `https://testuser:password@recipes.example.com` and have Authelia pick that up?

Forwarding the Cookie, or *Authorization headers, or sending its own Proxy-Authorization header with 1FA credentials using the basic scheme.

**Reply by cinderblockgames** [reply]
> How does it stink for your use case? It will always show as up if Authelia and the proxy are up, so there's no point in having the status check at all for individual apps, it's just an inefficient request. 

It would only redirect if the container is up, but otherwise, fair points all.  Thanks.

**Reply by james-d-elliott** [reply]
> It would only redirect if the container is up, but otherwise, fair points all. Thanks.

Hence why it should be configurable. One other aspect to this is the status check should allow the end user to determine which status codes are a successful response. For example a 30x should in your use case indicate the container is *most likely* up (Traefik does not necessarily do this status check unless you use the label and labels are not the only way to configure Traefik, in addition the status of a container may not change immediately and traefik may not pick up the container status change immediately) so they should not be followed. 

In addition a `405 Method Not Allowed` response would also indicate the container is running, and is responding correctly. This is because it indicates the path exists and can serve other method verbs, but the particular method verb used is not appropriate. This should only occur if the container is up and serving requests, and the endpoint may service other request method verbs. 405 is pretty unique in this way, though this argument could theoretically be argued for several other 4xx status codes depending on the application. Making it fairly important that this is something users can configure.

**Reply by james-d-elliott** [reply]
I'm going to adjust most of the static assets and the ~~`/api/status`~~ `/api/heath` endpoint to respond to HEAD requests as it's easier than I thought and the only issues I can see it introducing are when someone has misconfigured something (which will have people complain that it's a breaking change even though it just isn't) however I think all of those points still stand regarding appropriate status checks should have configurable request method, request headers, request body, valid response statuses, and choice of following redirects.

**Reply by cinderblockgames** [reply]
I agree, and thank you.
